client/core/bot.go

```diff
@@ -0,0 +1,1734 @@
+		delete(m.ords, oid)
+
+		switch {
+		case o.Filled == o.Qty:
+			m.log.Tracef("processTrade: order filled")
+		case o.Status == order.OrderStatusCanceled:
+			if len(o.Matches) == 0 {
+				m.log.Tracef("processTrade: order canceled WITHOUT matches")
+			} else {
+				m.log.Tracef("processTrade: order canceled WITH matches")
+			}
+		}
+		return
+	} else {
+		// Update our reference.
+		m.ords[oid] = o
+	}
+}
+
+// rebalance rebalances the makerBot's orders.
+//  1. Generate a basis price, p, adjusted for oracle weighting and bias.
+//  2. Apply the gap strategy to get a target spread, s.
+//  3. Check existing orders, if out of bounds
+//     [p +/- (s/2) - drift_tolerance, p +/- (s/2) + drift_tolerance],
+//     cancel the order
+//  4. Compare remaining order counts to configured, lots, and place new
+//     orders.
+func (m *makerBot) rebalance(ctx context.Context, newEpoch uint64) {
+	if !atomic.CompareAndSwapUint32(&m.rebalanceRunning, 0, 1) {
+		return
+	}
+	defer atomic.StoreUint32(&m.rebalanceRunning, 0)
+
+	newBuyLots, newSellLots, buyPrice, sellPrice := rebalance(ctx, m, m.market, m.program(), m.log, newEpoch)
+
+	// Place buy orders.
+	if newBuyLots > 0 {
+		ord := m.placeOrder(uint64(newBuyLots), buyPrice, false)
+		if ord != nil {
+			var oid order.OrderID
+			copy(oid[:], ord.ID)
+			m.ordMtx.Lock()
+			m.ords[oid] = ord
+			m.ordMtx.Unlock()
+		}
+	}
+
+	// Place sell orders.
+	if newSellLots > 0 {
+		ord := m.placeOrder(uint64(newSellLots), sellPrice, true)
+		if ord != nil {
+			var oid order.OrderID
+			copy(oid[:], ord.ID)
+			m.ordMtx.Lock()
+			m.ords[oid] = ord
+			m.ordMtx.Unlock()
+		}
+	}
+}
+
+// rebalancer is a stub to enabling testing of the rebalance calculations.
+type rebalancer interface {
+	basisPrice() uint64
+	halfSpread(basisPrice uint64) (uint64, error)
+	sortedOrders() (buys, sells []*sortedOrder)
+	cancelOrder(oid order.OrderID) error
+	MaxBuy(host string, base, quote uint32, rate uint64) (*MaxOrderEstimate, error)
+	MaxSell(host string, base, quote uint32) (*MaxOrderEstimate, error)
+}
+
+func rebalance(ctx context.Context, m rebalancer, mkt *Market, pgm *MakerProgram, log dex.Logger, newEpoch uint64) (newBuyLots, newSellLots int, buyPrice, sellPrice uint64) {
+	basisPrice := m.basisPrice()
+	if basisPrice == 0 {
+		log.Errorf("No basis price available")
+		return
+	}
+
+	log.Tracef("rebalance: basis price = %d", basisPrice)
+
+	// Three of the strategies will use a break-even half-gap.
+	var breakEven uint64
+	switch pgm.GapStrategy {
+	case GapStrategyAbsolutePlus, GapStrategyPercentPlus, GapStrategyMultiplier:
+		var err error
+		breakEven, err = m.halfSpread(basisPrice)
+		if err != nil {
+			log.Errorf("Could not calculate break-even spread: %v", err)
+			return
+		}
+	}
+
+	// Apply the base strategy.
+	var halfSpread uint64
+	switch pgm.GapStrategy {
+	case GapStrategyMultiplier:
+		halfSpread = uint64(math.Round(float64(breakEven) * (1 + pgm.GapFactor)))
```

Comment:
Why is this `1 + pgm.GapFactor`? It seems correct if it is exactly break even when the `GapFactor` is 1.